modules/tinymce/src/themes/silver/main/ts/ui/menus/menu/searchable/SearchableMenuField.ts

```diff
@@ -0,0 +1,182 @@
+import { AddEventsBehaviour, AlloyComponent, AlloyEvents, AlloySpec, AlloyTriggers, Behaviour, Input, Keying, NativeEvents, NativeSimulatedEvent, Representing } from '@ephox/alloy';
+import { Optional } from '@ephox/katamari';
+import { Attribute, EventArgs, SelectorFind } from '@ephox/sugar';
+
+import { UiFactoryBackstageProviders } from '../../../../backstage/Backstage';
+import { selectableClass as usualItemClass } from '../../item/ItemClasses';
+import { redirectMenuItemInteractionEvent, RedirectMenuItemInteractionEventData, refetchTriggerEvent } from './SearchableMenuEvents';
+
+// This is not stored in ItemClasses, because the searcher is not actually
+// contained within items. It isn't part of their navigation, and it
+// isn't maintained by menus. It is just part of the first menu, but
+// not its items.
+const menuSearcherClass = 'tox-menu__searcher';
+
+export interface MenuSearcherState {
+  readonly fetchPattern: string;
+  readonly selectionStart: number;
+  readonly selectionEnd: number;
+}
+
+// Ideally, we'd be using mementos to find it again, but we'd need to pass
+// that memento onto the dropdown, which isn't going to have it. Especially,
+// because the dropdown isn't responsible for putting this searcher component
+// into the menu, NestedMenus is.
+export const findWithinSandbox = (sandboxComp: AlloyComponent): Optional<AlloyComponent> => {
+  return SelectorFind.descendant(sandboxComp.element, `.${menuSearcherClass}`).bind(
+    (inputElem) => sandboxComp.getSystem().getByDom(inputElem).toOptional()
+  );
+};
+
+// There is nothing sandbox-specific about this code. It just needs to be
+// a container that wraps the search field.
+export const findWithinMenu = findWithinSandbox;
+
+export const restoreState = (inputComp: AlloyComponent, searcherState: MenuSearcherState): void => {
+  Representing.setValue(inputComp, searcherState.fetchPattern);
+  inputComp.element.dom.selectionStart = searcherState.selectionStart;
+  inputComp.element.dom.selectionEnd = searcherState.selectionEnd;
+};
+
+export const saveState = (inputComp: AlloyComponent): MenuSearcherState => {
+  const fetchPattern = Representing.getValue(inputComp);
+  const selectionStart = inputComp.element.dom.selectionStart;
+  const selectionEnd = inputComp.element.dom.selectionEnd;
+  return {
+    fetchPattern,
+    selectionStart,
+    selectionEnd
+  };
+};
+
+export interface MenuSearcherSpec {
+  readonly placeholder: Optional<string>;
+  readonly i18n: UiFactoryBackstageProviders['translate'];
+}
+
+// Make sure there is ARIA communicating the currently active item in the results.
+export const setActiveDescendant = (inputComp: AlloyComponent, active: AlloyComponent): void => {
+  const id = Attribute.get(active.element, 'id');
+  // Only set the ARIA attribute if we found a valid id
+  if (id !== undefined) {
+    Attribute.set(inputComp.element, 'aria-activedescendant', id);
+  }
```

Comment:
Should this be using `Attribute.getOpt` instead?

Reply:
- [ ] use Attribute.getOpt